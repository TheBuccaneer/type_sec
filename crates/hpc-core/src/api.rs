//! Minimaler, auskommentierter Skeleton der künftigen Safe-API.
//! Greift nur auf die Zustandsmarker aus `buffer::state` zurück und
//! kollidiert nicht mit dem bestehenden `buffer`-Code.

#![allow(dead_code, unused_imports)]

use core::marker::PhantomData;
use crate::buffer::state::{State, Empty, Ready, InFlight};
// Falls der Typ in deinem Code anders heißt, passe das Alias hier an:
use crate::buffer::GpuBuffer as LlBuffer;

/// Root-Lebensraum. In echt: OpenCL-Context/Devices/Programme.
pub struct Context {
    _opaque: (),
}

/// Befehlswarteschlange, an `Context` gebunden.
/// Lifetime erzwingt: Queue lebt nicht länger als der Context.
pub struct Queue<'ctx> {
    _ctx: &'ctx Context,
}

/// Ein typisierter GPU-Buffer im Host-API-Layer.
/// 
/// # Must Use
/// Nicht stumm verwerfen – ein nicht verwendeter Buffer kann auf ein Leck hindeuten
/// oder einen nicht vollzogenen State-Hop.
#[must_use]
pub struct DeviceBuffer<T, S: State> {
    inner: LlBuffer<S>,
    len_elems: usize,
    _ty: PhantomData<T>,
}

/// Kernel-Handle, an Queue/Context gebunden.
pub struct Kernel<'ctx> {
    _q: PhantomData<&'ctx ()>,
}

/// Ein Event-Token als Resultat eines GPU-Befehls.
/// 
/// # Must Use
/// Muss von [`Queue::wait`] oder einem Äquivalent konsumiert werden.
/// Nicht verwendete Events bedeuten, dass ein Kommando evtl. nie abgeschlossen wurde.
#[must_use]
#[derive(Debug)]
pub struct EventToken<'ctx> {
    _q: PhantomData<&'ctx ()>,
}



// --- Signaturen (nur Skeleton) ------------------------------------------------
impl Context {
    #[cfg(feature = "api-dev")]
    pub fn new() -> Self { Self { _opaque: () } }

    #[allow(clippy::new_without_default)]
    #[cfg(not(feature = "api-dev"))]
    pub fn new() -> Self { unimplemented!() }

    #[cfg(feature = "api-dev")]
    pub fn queue(&self) -> Queue<'_> { Queue { _ctx: self } }

    #[cfg(not(feature = "api-dev"))]
    pub fn queue(&self) -> Queue<'_> { unimplemented!() }


    pub fn create_queue(&self) -> Queue<'_> {
        Queue { _ctx: self }
    }
    
}

impl<'ctx> Queue<'ctx> {

    /// Device → Host (blocking):
    /// Liest den Inhalt eines `DeviceBuffer<T, Ready>` in `out` und blockiert,
    /// bis der Transfer abgeschlossen ist. Der Buffer bleibt `Ready`.
    #[inline]
    pub fn read_blocking<T>(
        &'ctx self,
        _buf: &DeviceBuffer<T, Ready>,
        _out: &mut [T],
    ) {
        // später: low-level clEnqueueReadBuffer + clFinish/Wait
        // Größenprüfung: out.len() == buf.len_elems()
        unimplemented!("Queue::read_blocking<T>: typed blocking read from device")
    }

    #[inline]
    pub fn create_buffer_elems<T>(&'ctx self, _n: usize) -> DeviceBuffer<T, Empty> {
        unimplemented!("Queue::create_buffer_elems: allocate n * size_of::<T>() bytes and wrap")
    }

    /// Lege einen noch leeren Gerätepuffer an.
    pub fn create_buffer<T>(&'ctx self) -> DeviceBuffer<T, Empty> {
        unimplemented!()
    }

/// Host→Device Write: Empty -> Ready.
///
/// # Examples
/// ```no_run
/// use hpc_core::api::Context;
/// let ctx = Context::new();
/// let q = ctx.queue();
/// let buf = q.create_buffer_elems::<u32>(16);
/// let _ready = q.enqueue_write(buf, &[0u32; 16]);
/// ```
///
/// ```compile_fail
/// // Write auf bereits Ready → sollte nicht kompilieren.
/// use hpc_core::api::Context;
/// let ctx = Context::new();
/// let q = ctx.queue();
/// let ready = {
///     let b = q.create_buffer_elems::<u32>(16);
///     q.enqueue_write(b, &[0u32; 16])
/// };
/// let _illegal = q.enqueue_write(ready, &[1u32; 16]);
/// ```
/// 
#[inline]
#[cfg(feature = "api-dev")]
pub fn enqueue_write<T>(
    &'ctx self,
    buf: DeviceBuffer<T, Empty>,
    _data: &[T],
) -> DeviceBuffer<T, Ready> {
    // Dev-Stitching: nur State-Übergang, keine echte Kopie.
    let inner_empty = buf.into_inner();
    // Low-Level: State-Hop über Helper (kein transmute im API-Layer).
    let inner_ready = unsafe { inner_empty.assume_state::<Ready>() };
    DeviceBuffer::from_inner_unchecked(inner_ready)
}

#[inline]
#[cfg(not(feature = "api-dev"))]
pub fn enqueue_write<T>(
    &'ctx self,
    _buf: DeviceBuffer<T, Empty>,
    _data: &[T],
) -> DeviceBuffer<T, Ready> {
    unimplemented!("Queue::enqueue_write<T>: typed write to device buffer")
}

      // Ready -> InFlight + EventToken
    #[cfg(feature = "api-dev")]
    #[inline]
    pub fn enqueue_kernel<T>(
        &'ctx self,
        _k: &Kernel<'ctx>,
        buf: DeviceBuffer<T, Ready>,
    ) -> (DeviceBuffer<T, InFlight>, EventToken<'ctx>) {
        // reiner State-Übergang; ersetzt du später durch echten enqueue
        let inner_ready = buf.into_inner();
        let inner_inflight: LlBuffer<InFlight> = unsafe { inner_ready.assume_state() };
        (DeviceBuffer::from_inner_unchecked(inner_inflight), EventToken { _q: core::marker::PhantomData })
    }

        /// Enqueued einen Kernel und liefert ein [`EventToken`].
    ///
    /// # State-Hop
    /// Eingänge: *Pending* → nach erfolgreichem Lauf: *Ready*.
    ///
    /// # Bemerkung
    /// Diese Funktion **führt** den Kernel irgendwann aus (Scheduling),
    /// das eigentliche Warten passiert über [`Queue::wait`].
    #[cfg(not(feature = "api-dev"))]
    #[inline]
    pub fn enqueue_kernel<T>(
        &'ctx self,
        _k: &Kernel<'ctx>,
        _buf: DeviceBuffer<T, Ready>,
    ) -> (DeviceBuffer<T, InFlight>, EventToken<'ctx>) {
        unimplemented!("Queue::enqueue_kernel<T>: start kernel, return InFlight + EventToken")
    }

   // InFlight -> Ready (Token wird konsumiert)
    #[cfg(feature = "api-dev")]
    #[inline]
    pub fn wait<T>(
        &'ctx self,
        _ev: EventToken<'ctx>,
        buf: DeviceBuffer<T, InFlight>,
    ) -> DeviceBuffer<T, Ready> {
        let inner_inflight = buf.into_inner();
        let inner_ready: LlBuffer<Ready> = unsafe { inner_inflight.assume_state() };
        DeviceBuffer::from_inner_unchecked(inner_ready)
    }

    /// Wartet/blockiert auf ein Event und vollzieht den State-Hop.
    ///
    /// # State-Hop
    /// *Pending* → *Ready*. Auf bereits *Ready* sollte **nicht** gewartet werden
    /// (siehe compile-fail-Tests).
    #[cfg(not(feature = "api-dev"))]
    #[inline]
    pub fn wait<T>(
        &'ctx self,
        _ev: EventToken<'ctx>,
        _buf: DeviceBuffer<T, InFlight>,
    ) -> DeviceBuffer<T, Ready> {
        unimplemented!("Queue::wait<T>: block on event, transition InFlight->Ready")
    }
}

impl<'ctx> Kernel<'ctx> {
    // dev: simpler Platzhalter
    #[cfg(feature = "api-dev")]
    pub fn new(_q: &'ctx Queue<'ctx>, _name: &str) -> Self {
        Self { _q: core::marker::PhantomData }
    }

        /// Erstellt ein Kernel-Handle für die gegebene Queue und Kernel-Bezeichnung.
    ///
    /// # Zweck
    /// *Nur Handle, kein Dispatch.* Der eigentliche Start passiert über
    /// [`Queue::enqueue_kernel`].
    ///
    /// # State-Hop
    /// Vorbereitung -> ausführbar mit `enqueue_kernel`.
    #[inline]
    #[cfg(not(feature = "api-dev"))]
    pub fn new(_q: &'ctx Queue<'ctx>, _name: &str) -> Self { unimplemented!() }
}


impl<T, S: State> core::fmt::Debug for DeviceBuffer<T, S> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        f.debug_struct("DeviceBuffer")
            .field("len_elems", &self.len_elems)
            .finish_non_exhaustive()
    }
}

impl<T, S: State> DeviceBuffer<T, S> {
    /// Interne Konstruktion aus Low-Level-Buffer + Elementlänge.
    /// (Sichtbar im Crate; die High-Level-APIs stellen das später sicher her.)
    pub(crate) fn from_inner(inner: LlBuffer<S>, len_elems: usize) -> Self {
        Self { inner, len_elems, _ty: PhantomData }
    }

    /// Gibt den Low-Level-Buffer wieder zurück (z. B. für interne Delegation).
    pub(crate) fn into_inner(self) -> LlBuffer<S> {
        self.inner
    }

    /// Anzahl Elemente (nicht Bytes).
    #[inline] pub fn len_elems(&self) -> usize { self.len_elems }

    /// Byte-Länge, abgeleitet aus `T`.
    #[inline] pub fn len_bytes(&self) -> usize { self.len_elems * core::mem::size_of::<T>() }

    pub(crate) fn from_inner_unchecked(inner: LlBuffer<S>) -> Self {
    let bytes = inner.len_bytes();
    let sz = core::mem::size_of::<T>();
    debug_assert!(
        bytes % sz == 0,
        "DeviceBuffer<{}>: bytes ({}) nicht durch size_of T ({}) teilbar",
        core::any::type_name::<T>(),
        bytes,
        sz
    );
    let len_elems = bytes / sz;
    Self { inner, len_elems, _ty: PhantomData }
}

}
