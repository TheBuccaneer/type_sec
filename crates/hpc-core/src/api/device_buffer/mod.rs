// src/api/device_buffer/mod.rs

use crate::buffer::{GpuBuffer, state::State};
use std::marker::PhantomData;

// State-specific implementations
mod empty;
mod inflight;
mod ready;

//=============================================================================
// STRUCT DEFINITION
//=============================================================================

/// High-Level typisierter Buffer. Delegiert an den GPUBuffer
///
/// User API: Buffers können erstellt, gelesen, geschrieben und für Kernels genutzt werden.
/// Die interne Struktur ist versteckt - Type-State Pattern sorgt für Compile-Time Safety.
#[derive(Debug)]
pub struct DeviceBuffer<'brand, T, S: State> {
    pub(crate) inner: GpuBuffer<S>,
    pub(crate) len: usize,
    // Zwei separate Marker:
    pub(crate) _brand: PhantomData<fn(&'brand ()) -> &'brand ()>, // invariant für Context
    pub(crate) _type: PhantomData<T>, // kovariant für Typ (das ist OK)
}

impl<'ctx, T, S: State> DeviceBuffer<'ctx, T, S> {
    pub(crate) fn from_inner(inner: GpuBuffer<S>, len_elems: usize) -> Self {
        Self {
            inner,
            len: len_elems,
            _brand: PhantomData,  // ← für Context-Branding
            _type: PhantomData,   // ← für Typ-Parameter
        }
    }
}