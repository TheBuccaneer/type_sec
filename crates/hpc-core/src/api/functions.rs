// src/api/functions.rs

use crate::buffer::{GpuBuffer, state::Empty};
use crate::error::Result;
use super::{Context, DeviceBuffer};

//=============================================================================
// BUFFER CREATION
//=============================================================================

/// Erstellt einen neuen Buffer auf Nutzerebene!
/// 
/// # Beispiel
/// ```rust
/// let (ctx, queue) = Context::new_first_gpu()?;
/// let buffer = create_buffer::<f32>(&ctx, 1000)?;  // 1000 f32 Elemente
/// ```
pub fn create_buffer<T>(ctx: &Context, n_elems: usize) -> Result<DeviceBuffer<'static, T, Empty>> {
    let inner: GpuBuffer<Empty> = GpuBuffer::<Empty>::create_uninit_elems::<T>(ctx.raw(), n_elems)?;
    Ok(DeviceBuffer::from_inner(inner, n_elems))
}