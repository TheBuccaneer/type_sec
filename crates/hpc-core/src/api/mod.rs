// src/api/mod.rs

// Re-export von Error types
pub use crate::error::{Error, Result};

// Submodule
mod device_buffer;
mod opencl;
mod functions;

// Re-exports der Submodule
pub use device_buffer::DeviceBuffer;
pub use opencl::{Context, Queue, Kernel};
pub use functions::create_buffer;

mod util;
pub use util::{EventToken, ReadGuard};