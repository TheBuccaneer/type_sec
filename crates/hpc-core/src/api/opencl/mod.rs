// src/api/opencl/mod.rs

mod context;
mod kernel;
mod queue;

pub use context::Context;
pub use kernel::Kernel;
pub use queue::Queue;
