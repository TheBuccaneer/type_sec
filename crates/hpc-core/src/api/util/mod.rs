// src/api/util/mod.rs

mod event_token;
mod read_guard;

pub use event_token::EventToken;
pub use read_guard::ReadGuard;
