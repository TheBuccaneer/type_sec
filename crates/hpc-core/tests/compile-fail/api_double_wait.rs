// Cargo macht aus "hpc-core" -> "hpc_core"
use hpc_core::api::{Context, Kernel};

fn main() {
    let ctx = Context::new();
    let q = ctx.queue();

    // typisierte Allokation + erster Write (Bodies sind egal, trybuild prüft nur Typen)
    let buf = q.create_buffer_elems::<u32>(16);
    let ready = q.enqueue_write(buf, &[0u32; 16]);

    let k = Kernel::new(&q, "dummy");

    // 1. Kernel → InFlight + Event
    let (inflight, ev) = q.enqueue_kernel(&k, ready);

    // 1. wait() verbraucht `ev`
    let ready_again = q.wait(ev, inflight);

    // 2. Kernel → neuer Event (ev2)
    let (inflight2, _ev2) = q.enqueue_kernel(&k, ready_again);

    // ❌ compile-fail: `ev` wurde bereits bewegt (use of moved value)
    let _illegal = q.wait(ev, inflight2);
}
