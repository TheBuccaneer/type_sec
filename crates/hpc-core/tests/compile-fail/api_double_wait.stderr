error[E0382]: use of moved value: `ev`
 --> tests/compile-fail/api_double_wait.rs
  |
  |     let (inflight, ev) = q.enqueue_kernel(&k, ready);
  |                    -- move occurs because `ev` has type `EventToken<'_>`, which does not implement the `Copy` trait
...
  |     let ready_again = q.wait(ev, inflight);
  |                              -- value moved here
...
  |     let _illegal = q.wait(ev, inflight2);
  |                           ^^ value used here after move
